breakpad-symbols/src/sym_file/parser.rs

```diff
@@ -27,265 +35,282 @@ enum Line {
     StackCfi(StackInfoCfi),
 }
 
-// Nom's `eol` doesn't use complete! so it will return Incomplete.
-named!(
-    my_eol<char>,
-    complete!(preceded!(many0!(char!('\r')), char!('\n')))
-);
+/// Match a hex string, parse it to a u64.
+fn hex_str_u64(input: &[u8]) -> IResult<&[u8], u64> {
+    map_res(map_res(hex_digit1, str::from_utf8), |s| {
+        u64::from_str_radix(s, 16)
+    })(input)
+}
+
+/// Match a decimal string, parse it to a u32.
+fn decimal_u32(input: &[u8]) -> IResult<&[u8], u32> {
+    map_res(map_res(digit1, str::from_utf8), FromStr::from_str)(input)
+}
 
-// Match a hex string, parse it t)o a u64.
-named!(hex_str_u64<&[u8], u64>,
-       map_res!(map_res!(hex_digit, str::from_utf8), |s| u64::from_str_radix(s, 16)));
+/// Take 0 or more non-space bytes.
+fn non_space(input: &[u8]) -> IResult<&[u8], &[u8]> {
+    take_while(|c: u8| c != b' ')(input)
+}
+
+/// Accept `\n` with an arbitrary number of preceding `\r` bytes.
+///
+/// This is different from `line_ending` which doesn't accept `\r` if it isn't
+/// followed by `\n`.
+fn my_eol(input: &[u8]) -> IResult<&[u8], char> {
+    preceded(many0(char('\r')), char('\n'))(input)
+}
 
-// Match a decimal string, parse it to a u32.
-named!(decimal_u32<&[u8], u32>, map_res!(map_res!(digit, str::from_utf8), FromStr::from_str));
+/// Accept everything except `\r` and `\n`.
+///
+/// This is different from `not_line_ending` which rejects its input if it's
+/// followed by a `\r` which is not immediately followed by a `\n`.
+fn not_my_eol(input: &[u8]) -> IResult<&[u8], &[u8]> {
+    take_while(|b| b != b'\r' && b != b'\n')(input)
+}
+
+/// Parse a single byte if it matches the predicate.
+///
+/// nom has `satisfy`, which is similar. It differs in the argument type of the
+/// predicate: `satisfy`'s predicate takes `char`, whereas `single`'s predicate
+/// takes `u8`.
+fn single(predicate: fn(u8) -> bool) -> impl Fn(&[u8]) -> IResult<&[u8], u8> {
+    move |i: &[u8]| match i.split_first() {
+        Some((b, rest)) if predicate(*b) => Ok((rest, *b)),
+        _ => Err(Err::Error(Error::from_error_kind(i, ErrorKind::Satisfy))),
+    }
+}
 
 // Matches a MODULE record.
-named!(module_line<&[u8], ()>,
-  chain!(
-    tag!("MODULE") ~
-          space     ~
-          // os
-    alphanumeric ~
-          space ~
-          // cpu
-    take_until!(" ") ~
-          space ~
-          // debug id
-    hex_digit ~
-          space ~
-          // filename
-    not_line_ending ~
-    my_eol ,
-    || {}
-));
+fn module_line(input: &[u8]) -> IResult<&[u8], ()> {
+    let (input, _) = tuple((
+        terminated(tag("MODULE"), multispace1),
+        terminated(non_space, multispace1),  // os
+        terminated(non_space, multispace1),  // cpu
+        terminated(hex_digit1, multispace1), // debug id
+        terminated(not_my_eol, my_eol),      // filename
+    ))(input)?;
+    Ok((input, ()))
```

Comment:
huh, interesting that we discard this, I never even registered that these fields existed in a symbol file.

Reply:
I guess it could be nice to get the debug ID and filename out to validate that they match what you expect, but AFAIK nothing has ever done that and it hasn't seemed to matter. IME, the other bits are mostly useful just for manually spot-checking symbol files.